docs/architecture/dapr-for-net-developers/publish-subscribe-buildingblock.md

```diff
@@ -0,0 +1,398 @@
+
+Below you see an example of a Dapr configuration file for configuring a RabbitMQ message broker component:
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: pubsub-rq
+spec:
+  type: pubsub.rabbitmq
+  metadata:
+  - name: host
+    value: "amqp://localhost:5672"
+  - name: durable
+    value: true
+```
+
+In this example, you can see that you can specify any message broker-specific configuration in the `metadata` block. In this case, RabbitMQ is configured to create durable queues. Each of the components' configuration will have its own set of possible fields. You can read which fields are available in the documentation of each [Publish/Subscribe component](https://github.com/dapr/components-contrib/tree/master/pubsub).
+
+The Publish/Subscribe building block also provides a declarative way of subscribing to a topic. Below you see an example of a Dapr configuration file for configuring a subscription:
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Subscription
+metadata:
+  name: newOrder-subscription
+spec:
+  pubsubname: pubsub
+  topic: newOrder
+  route: /orders
+scopes:
+- ServiceB
+- ServiceC
+```
+
+You have to specify several elements with every subscription:
+
+- The name of the Dapr Publish/Subscribe component you want to use (in this case `pubsub`).
+- The name of the topic to subscribe to (in this case `newOrder`).
+- The API operation that needs to be called for this topic (in this case `/orders`).
+- The [scope](https://docs.dapr.io/developing-applications/building-blocks/pubsub/pubsub-scopes/) can specify which services can publish and subscribe to a topic.
+
+## Reference architecture: eShopOnDapr
+
+As mentioned earlier in an earlier chapter, the accompanying [eShopOnDapr](https://github.com/dotnet-architecture/eShopOnDapr) app provides an end-to-end reference architecture for constructing a microservices application implementing Dapr. eShopOnDapr is an evolution of the widely popular [eShopOnContainers](https://github.com/dotnet-architecture/eShopOnContainer) app, created several years ago. Both versions use the Publish/Subscribe pattern for communicating [integration events](https://devblogs.microsoft.com/cesardelatorre/domain-events-vs-integration-events-in-domain-driven-design-and-microservices-architectures/#integration-events) across microservices. Integration events include:
+
+- When a user checks-out a shopping basket.
+- When a payment for an order has succeeded.
+- When the grace-period of a purchase has expired.
+
+Eventing in eShopOnContainers is based on the following `IEventBus` interface:
+
+```csharp
+public interface IEventBus
+{
+    void Publish(IntegrationEvent @event);
+
+    void Subscribe<T, TH>()
+        where T : IntegrationEvent
+        where TH : IIntegrationEventHandler<T>;
```

Comment:
```suggestion
    void Subscribe<T, THandler>()
        where T : IntegrationEvent
        where THandler : IIntegrationEventHandler<T>;
```

As a C# developer myself, I typically prefer when there are multiple type parameters that at the very least the second parameter has a bit more explicit name. I'd even argue that when you have two, the first could also benefit from that. You might even consider:

```csharp
void Subscribe<TEvent, THandler>()
        where TEvent : IntegrationEvent
        where THandler : IIntegrationEventHandler<TEvent>;
```

While a bit more verbose, it is much much easier to read.

Reply:
I totally agree. This code was copied as-is from the eShopOnContainers codebase, but for the reader of the book, it is better to be more verbose. 
We will also update the code in eShopOnDapr to use more verbose type parameters and reflect that in the sample snippets.